osf/models/node.py

```diff
@@ -135,29 +136,28 @@ def can_view(self, user=None, private_link=None):
             qs |= self.filter(private_links__is_deleted=False, private_links__key=private_link)
 
         if user is not None and not isinstance(user, AnonymousUser):
-            if isinstance(user, OSFUser):
-                user = user.pk
-            if not isinstance(user, int):
-                raise TypeError('"user" must be either {} or {}. Got {!r}'.format(int, OSFUser, user))
-
-            sqs = Contributor.objects.filter(node=models.OuterRef('pk'), user__id=user, read=True)
-            qs |= self.annotate(can_view=models.Exists(sqs)).filter(can_view=True)
+            read_user_query = get_objects_for_user(user, 'read_node', self)
+            qs |= read_user_query
             qs |= self.extra(where=["""
                 "osf_abstractnode".id in (
                     WITH RECURSIVE implicit_read AS (
-                        SELECT "osf_contributor"."node_id"
-                        FROM "osf_contributor"
-                        WHERE "osf_contributor"."user_id" = %s
-                        AND "osf_contributor"."admin" is TRUE
+                        SELECT DISTINCT N.id as node_id
+                        FROM osf_abstractnode as N, auth_permission as P, guardian_groupobjectpermission as G, django_content_type as CT, osf_osfuser_groups as UG
+                        WHERE P.codename = 'admin_node'
+                        AND G.permission_id = P.id
+                        AND CT.model = 'abstractnode'
+                        AND UG.osfuser_id = %s
+                        AND G.group_id = UG.group_id
+                        AND G.object_pk::integer = N.id
+                        AND N.type = 'osf.node'
                     UNION ALL
                         SELECT "osf_noderelation"."child_id"
                         FROM "implicit_read"
                         LEFT JOIN "osf_noderelation" ON "osf_noderelation"."parent_id" = "implicit_read"."node_id"
                         WHERE "osf_noderelation"."is_node_link" IS FALSE
                     ) SELECT * FROM implicit_read
                 )
-            """], params=(user, ))
-
+            """], params=(user.id, ))
         return qs
```

Comment:
@caseyrollins, I think I got this query working with OSF Groups!
If you do Node.objects.can_view({user}), it will return all the nodes you have permission to read, plus any nodes that you have implicit admins on the parent.  You could have gotten this permission from being explicitly added to the node as a contributor, or through your OSF Group. [Test here ](https://github.com/pattisdr/osf.io/pull/11/commits/4b2a533772399528339320a0206f8f72a105185a#diff-37f6f65b17e6936095c8c901ca02c8f7R276)